storage/pruner/pruner.go

```diff
@@ -0,0 +1,374 @@
+)
+
+type PruneMethod int
+
+const (
+	// PruneByHeight prunes data using finalized block heights
+	// This is should be used when a db does not include a view index
+	PruneByHeight PruneMethod = iota + 1
+
+	// PruneByView prunes data using views
+	// This is the simpler and more efficient method, but requires the node to have a view index
+	PruneByView
+
+	// DefaultPruneInterval is the default interval between pruning cycles
+	DefaultPruneInterval = 5 * time.Minute
+
+	// DefaultThrottleDelay is the default delay between pruning individual blocks
+	DefaultThrottleDelay = 20 * time.Millisecond
+)
+
+type Pruner struct {
+	component.Component
+
+	log zerolog.Logger
+	db  *pebble.DB
+
+	headers        storage.Headers
+	chunkDataPacks storage.ChunkDataPacks
+
+	targetKeepBlocks uint64 // number of blocks to keep below the highest block
+	pruneThreshold   uint64 // number of blocks to exceed the target before pruning
+	pruneMethod      PruneMethod
+
+	lowestHeader     *flow.Header
+	rootSealedHeader *flow.Header
+	progress         storage.ConsumerProgress
+
+	producers []*BlockProcessedProducer
+	mu        sync.RWMutex
+}
+
+func New(
+	log zerolog.Logger,
+	db *pebble.DB,
+	headers storage.Headers,
+	chunkDataPacks storage.ChunkDataPacks,
+	rootSealedHeader *flow.Header,
+	progress storage.ConsumerProgress,
+	targetKeepBlocks uint64,
+	pruneThreshold uint64,
+) *Pruner {
+	pruneMethod := PruneByHeight
+	if progress.Consumer() == module.ConsumeProgressProtocolDBPrunerByView {
+		pruneMethod = PruneByView
+	}
+
+	p := &Pruner{
+		log: log.With().Str("component", "protocoldb_pruner").Logger(),
+		db:  db,
+
+		headers:          headers,
+		chunkDataPacks:   chunkDataPacks,
+		targetKeepBlocks: targetKeepBlocks,
+		pruneThreshold:   pruneThreshold,
+		rootSealedHeader: rootSealedHeader,
+
+		progress:    progress,
+		pruneMethod: pruneMethod,
+	}
+
+	p.Component = component.NewComponentManagerBuilder().
+		AddWorker(p.loop).
+		Build()
+
+	return p
+}
+
+// Register registers an external component as a producer of BlockProcessed events.
+// The pruner will take the lowest processed height from the set of producers, and use that as the
+// base height for pruning.
+func (p *Pruner) Register() *BlockProcessedProducer {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// Note: the initial height is 0 so pruning will not start until all producers have completed
+	// processing at least one block
+	producer := NewBlockProcessedProducer()
+	p.producers = append(p.producers, producer)
+
+	return producer
+}
+
+// baseHeight returns the highest height that all block processors have completed processing
+// for instance, if consensus is at 1000, and execution is at 750, the base height should be 750
+// this ensures all subsystems have the data they need regardless of the pruning thresholds used
+func (p *Pruner) baseHeight() uint64 {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	lowest := uint64(0)
+	for _, producer := range p.producers {
```

Comment:
I feel `producers` is a bit too abstract. 

I think the producers are basically two producers: executedBlockProducer, and sealedBlockProducer. And the lowest of thembasically means the HighestSealedAndExecutedBlockHeight. Right?

I would suggest to decouple them:
We could have a module responsible for returning HighestSealedAndExecutedHeight, so those logic can be moved out of the pruning core and tested separately. And this module could probably be reused, as we might have other systems depending on highest sealed and executed height.

Reply:
I left it abstract on purpose because I think we will have different requirements for each node type, and possible different for various configurations. I'll need to vet this some more.

For the chunk data packs, we want to use the lower of sealed or executed as the base.
For SN, it's likely just sealed. For ANs, it's probably indexed and sealed.

There's an identical situation for the execution data db, were we want to prune based on the execution data requester engine or the indexer if it's enabled. So I agree that breaking this tracker out into a separate module that aggregates the heights into a single value is a good idea.